File: ch6/01_breadth_first_search.py
from collections import deque

def person_is_seller(name: str) -> bool:
    return name[-1] == 'm';

def search(graph: map, name: str) -> bool:
    search_queue = deque()
    search_queue += graph[name]
    searched = set()

    while search_queue:
        person = search_queue.popleft()
        print("- searching {}".format(person))

        if not person in searched:
            if person_is_seller(person):
                print("{} is a mango seller!".format(person))
                return True
            else:
                search_queue += graph[person]
                searched.add(person)

    print("no-one is a mango seller!")
    return False

graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

search(graph, "you")

File: ch6/01_breadth_first_search.php
<?php
class MyQueue extends SplQueue
{
    public function enqueueMany(string ...$values): void
    {
        foreach ($values as $value) {
            $this->enqueue($value);
        }
    }
}

function person_is_seller(string $name): bool {
    return $name[-1] === 'm';
}

function search(array $graph, string $name): bool
{
    $queue = new MyQueue;
    $queue->enqueueMany(...$graph[$name]);
    $searched = [];

    foreach ($queue as $person) {
        if (in_array($person, $searched)) {
            continue;
        }

        echo sprintf('- searched %s', $person) . "\n";

        if (person_is_seller($person)) {
            echo sprintf('%s is a mango seller!', $person) . "\n";
            return true;
        } else {
            $queue->enqueueMany(...$graph[$person]);
            $searched[] = $person;
        }
    }

    echo 'no-one is a mango seller' . "\n";
    return false;
}


$graph = [
    'you'    => ['alice', 'bob', 'claire'],
    'bob'    => ['anuj', 'peggy'],
    'alice'  => ['peggy'],
    'claire' => ['thom', 'jonny'],
    'anuj'   => [],
    'peggy'  => [],
    'thom'   => [],
    'jonny'  => [],
];

search($graph, 'you');
